[Controller] Zabbix controller initial commit

jimmys01 opened the pull request.

jimmys01 committed:
[Controller] Zabbix controller initial commit

jimmys01 commented:
I have an issue, the controller does not send the data when they are collected but store them. They are sent when the reboot command is given, I don't know why. Otherwise works fine. Datacenter guys will love this controller, I know I will.

jimmys01 committed:
Scheduler fix

I knew it had to bee something small

jimmys01 commented:
If anybody want to give it a try.
[firmware.zip](https://github.com/letscontrolit/ESPEasy/files/3325715/firmware.zip)

TD-er commented on src/_C015.ino:
```diff
@@ -0,0 +1,127 @@
+  client.write(packet_header, sizeof(packet_header) - 1);
+  client.write((char *)&content_len, sizeof(content_len));
+  client.write(packet_content, content_len);
+  client.stop(); 
```
Is this intentional, to stop the client here (in the for loop) ?
Also the indents are gone here, which makes it hard to read.

jimmys01 replied:
Yes, `client.flush` would not work, I spent 2 hours on this. I know that this is unificient

uzi18 replied:
maybe client.flush() ?

TD-er replied:
See [here](https://www.zabbix.org/wiki/Docs/protocols/zabbix_sender/3.4) how you can combine multiple values at once

jimmys01 replied:
I saw that, but I also got warried that the buffer that JSON would need to allocate would be a lot..

jimmys01 commented:
![εικόνα](https://user-images.githubusercontent.com/17748323/60110178-58016e00-9774-11e9-9d8f-6d51ce262c62.png)
![εικόνα](https://user-images.githubusercontent.com/17748323/60110225-6cde0180-9774-11e9-86ec-ffb62b96d48b.png)

:-)

TD-er commented on src/_C015.ino:
```diff
@@ -0,0 +1,127 @@
+    {
+
+
+const size_t capacity = JSON_ARRAY_SIZE(1) + JSON_OBJECT_SIZE(2) + JSON_OBJECT_SIZE(3);
```
Where does this come from?
Looks like this is something determined at compile time.
Ho much is it? (value is used to allocate an array on the stack)

jimmys01 replied:
this is the JSON library prefered way. Because in this PR I still use the 5.x arduJSON, i dont know if in the newer version they have a diferent way.

TD-er commented on src/_C015.ino:
```diff
@@ -0,0 +1,127 @@
+  char packet_content[capacity];
+  
+
+  root.printTo((char*)packet_content, root.measureLength()+1);
```
This is not taking into account the true size of the buffer `packet_content` you just created.

jimmys01 replied:
because we need the /0/0/0/0 this leaves at the end.

TD-er replied:
What I wanted to stress is that you don't check for the length of the buffer, but just print based on the size of the message.
Also the buffer is not initialized to 0's, which can be done like this:
```c++
char packet_content[capacity] = {0} ;
```

TD-er commented on src/_C015.ino:
```diff
@@ -0,0 +1,127 @@
+  
+
+  root.printTo((char*)packet_content, root.measureLength()+1);
+  unsigned long long content_len = sizeof(packet_content);
```
This is constant, since you're looking at the size of the buffer, not how much data you put into it.

jimmys01 replied:
I am not sure what you want to say, but `sizeof(packet_content)` returns the number of bytes until it reacheces a \0 if i recall correcty. So this is dependant on the JSON length

TD-er replied:
Hmm, you're right, I thought I read `size_of`.
Maybe 34+ degree is not that  well for my focus ;)

TD-er replied:
I've looked for it.
[sizeof](https://www.arduino.cc/en/pmwiki.php?n=Reference/Sizeof) does  return the size of the object, not the length of the string that may be stored in it.

TD-er commented on src/_C015.ino:
```diff
@@ -0,0 +1,127 @@
+  //addLog(LOG_LEVEL_INFO, String(F("ZBX: ")) + packet_content);
+  ControllerSettings.connectToHost(client);
+  client.write(packet_header, sizeof(packet_header) - 1);
+  client.write((char *)&content_len, sizeof(content_len));
```
I don't get this line.
`content_len` is an integer value (of 8 bytes size).
The result of `size_of` is very likely of type `size_t`, so why use an `uint64_t` type here?
And why print the content of that integer value in some binary format? (Endianness can also mess up what you try to send)

jimmys01 replied:
The protocol wants it in 4 bytes (long long) (https://www.zabbix.com/documentation/4.2/manual/appendix/protocols/header_datalen). Int is 2 bytes.

TD-er replied:
long long is 64 bit.
A long is 32 bit and I am quite sure int is also 32 bit.

When in doubt , use uint32_t.

jimmys01 replied:
uint32_t is not working for me, uint64_t works, even thought the protocol asks for 4 bytes.

jimmys01 commented:
Also keep in mind that if JSON 6.x is merged this code will need to be updated to it. Now it follows JSON 5.x

jimmys01 commented:
Any sugestions about the travis errors?

TD-er commented:
Well, you should also rename the C015 stuff to C017.
See this PR: https://github.com/letscontrolit/ESPEasy/pull/2430

TD-er commented:
And about the Travis errors.
You should uncomment this (C015 is right above it, but like I said you should use C017)
https://github.com/letscontrolit/ESPEasy/blob/b41c8a59d76cd4918d8444712668159079235006/src/_CPlugin_Helper.h#L463-L467

jimmys01 committed:
Zabbix secont attempt

jimmys01 committed:
Improvments

Opted as TDer sugested to send one (larger) packet, than openning and closing the connection for each value.

TD-er commented on src/_C017.ino:
```diff
@@ -84,44 +90,58 @@ bool do_process_c017_delay_queue(int controller_number, const C017_queue_element
+  data_0["value"] = atof(element.txt[0].c_str());            // ESPeasy supports only floats
 
-JsonArray& data = root.createNestedArray("data");
+  if (valueCount >= 1)
```
Should be >, not >=

Can't you use a for loop for this? You don't keep the objects outside the scope of the if-checks, so I don't see why you need to have these extra if's
The max. number of values is defined as MAX_VARS (have to check in the globals file)

jimmys01 replied:
Using loops to populate JSON is not straightforward. If I find time I will give it a try.
If the task has only 1 Value (valueCount will be 0) then the if statments wont run and the JSON wont populate any further. Anyway I am testing this 2 hours and it works (I am using Uptime, System Load, Free Heap and Free Stack as task values and also Temp and Humidity. Values are send every 60 sec.

TD-er replied:
Valuecount should not be 0 for a plugin that does deliver data.
If it does, then we have a bug.

jimmys01 replied:
All the other controllers have lines similar to this
```
        for (byte x = 0; x < valueCount; x++)
        {element.txt[x] = formatUserVarNoCheck(event, x);}
```
This tells me that x starts at 0 

I do the same essentialy check but in a opposite way (I check for >= instead of <)

TD-er replied:
Nope, x does start at 0, which is used as index in arrays.
So if `valueCount == 4`, `x` will loop from 0 ... 3
If valueCount = 0, then no occurrence of x will match, since `0 < 0` yields false.

This means any plugin that does output data (e.g. a display does not output values) should have at least 1 as valueCount.
N.B. 0 is also possible. (for the display plugin for example)

So your check will add `data_0` and `data_1` for a plugin that has 1 output value.

jimmys01 replied:
Ok I will fix it

jimmys01 replied:
So after digging arrount getValueCountFromSensorType() returns 4 even if only 1 taskvalue is used.
Example use "Generic - System Info " and use only one value for output configuration and leve the rest blank. The getValueCountFromSensorType() will return 4 instead of 1. I dont know if this is meant to be like that.

TD-er replied:
Depends on whether you did set the number of values and saved it (not just adding the plugin, alter it and save)

jimmys01 replied:
Would it make sence to ignore values whos value name ="" ??
For this condroller case an empty key does not seem to matter to Zabbix.

EG:
`{"request":"sender data","data":[{"host":"zabbix1","key":"uptime","value":13},{"host":"zabbix1","key":"freeheap","value":20160},{"host":"zabbix1","key":"rssi","value":-72},{"host":"zabbix1","key":"","value":0}]}`
The last part has an empty key

jimmys01 replied:
Thanks TD-er for the suggestions, I figured out the JSON for loop, after messing around with vectors and arrays... I will leave it overnight reporting every 30sec to the zabbix server, see if it still works in the morning

TD-er replied:
About ignoring values without names... There is already a feature request (can't find it now)  about adding some filter to controllers to select when to ignore values.
One thing could be when names are empty (or use a generated one like `name#1` ) or when a value is NaN.
Maybe for some controllers there can be more reasons to ignore (negative values perhaps?)

So I think we should first add a checkmark option to the controller settings for all controllers and then add some options when to consider "illegal" or "undesired" values. (or maybe let the controller decide which one is applicable)

jimmys01 committed:
More improvments

TD-er commented:
One more suggestion. Can you be more descriptive in the commit messages?
These will appear in the change log (or I must stash a PR into one new commit to the 'mega' branch)

jimmys01 commented:
Sure, I know I do a lot of small commits and some of them are hasty, but my time is really scarse and I worked in some of the code a lot more that I would like to admit. Of cource feel free to change those PRs as you see fit. Of cource I will be more descriptive from now on

jimmys01 commented:
![εικόνα](https://user-images.githubusercontent.com/17748323/60249150-d7a25080-98cc-11e9-81ba-fe22740f8152.png)
In the mean time the controller works fine. Whats that step in the Free Heap graph?

TD-er commented:
I am not sure about that jump in heap usage.
Can you see if that's about the same time a new WiFi connection has been made?
Maybe something is not cleaned up well? (not calling delete on something that was created using new)

jimmys01 commented:
> Connected: | 21h03m
> Number Reconnects: | 0

No reconnects no reboots.

jimmys01 commented:
The PR for the JSON6+ version of the Controller: jimmys01/ESPEasy#3
Should i also push it or I will make a mess?

TD-er commented:
What do you mean?
Are you now somewhat stuck until I merge something?
If so, what PR should be merged first?

jimmys01 commented:
If I push the updated Controller i will have also to push the JSON update, or else it will fail the Travis test. I will push the updated pr to this pr when you commit #2489 
Should I also include this Controller to the Normal build or leave it on Testing?

TD-er commented:
JSON PR has been merged, so you may rebase the branch if needed to the head of mega branch.
Leave the controller on testing for now.
And can you also increase the number of plugins? (it is now too low, see debug out on test build)

jimmys01 committed:
[ZABBIX] Update code to JSON6+

Also added F macros and a check for not sending empty TaskDeviceValueNames

jimmys01 committed:
Increace PLUGIN_MAX by 10

jimmys01 commented:
Is this going to be merged, I would like to promote it in some Zabbix forums.

TD-er commented:
There is still the remark I made here: https://github.com/letscontrolit/ESPEasy/pull/2495#discussion_r300488808

jimmys01 committed:
Couple payload_len to actual JSON payload length.

jimmys01 committed:
Resolve conflicts with C015

jimmys01 committed:
Resolving conflicts #2

jimmys01 committed:
Commit the changed define_plugin_sets.h after conflicts have been resolved

jimmys01 commented:
@TD-er it is up to you now, Github won't let me resolve this one line that is conflicting

TD-er committed:
Merge branch 'mega' into Zabbix_controller_push

TD-er commented:
@jimmys01 Done
Let Travis test it and then I will merge it later this evening.
I also have some commits to merge myself.

jimmys01 commented:
16 days uptime (had a power outage)
![εικόνα](https://user-images.githubusercontent.com/17748323/63522893-d4f55d80-c501-11e9-8bb0-0d779f89ea2c.png)
over a month of measurments
![εικόνα](https://user-images.githubusercontent.com/17748323/63523052-1c7be980-c502-11e9-911e-a3d28f1d82c9.png)

@TD-er The zabbix controller is working fine, perhaps its time to put it on a normal build?
Edit: this is one of my 3 test nodes.

TD-er commented:
We're running into build size issues lately, so I guess we may need to change what controllers and plugins are distributed in what build.
